csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -190,14 +190,470 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
                 // always unpin the input buffers, and delete the native Onnx value objects
                 for (int i = 0; i < inputs.Count; i++)
                 {
-                    NativeMethods.OrtReleaseValue(inputTensors[i]); // For elementary type Tensors, this should not release the buffer, but should delete the native tensor object.
-                                                                    // For string tensors, this releases the native memory allocated for the tensor, including the buffer
-                    pinnedBufferHandles[i].Dispose();
+                    NativeMethods.OrtReleaseValue(inputValuesArray[i]); // For elementary type Tensors, this should not release the buffer, but should delete the native tensor object.
+                                                                        // For string tensors, this releases the native memory allocated for the tensor, including the buffer
+                    pinnedInputBufferHandles[i].Dispose();
+                }
+            }
+
+        }
+
+        /// <summary>
+        /// Runs the loaded model for the given inputs, and fetches all the outputs.
+        /// </summary>
+        /// <param name="inputNames">Specify a collection of string that indicates the input names. Should match <paramref name="inputValues"/>.</param>
+        /// <param name="inputValues">Specify a collection of <see cref="PinnedOnnxValue"/> that indicates the input values.</param>
+        /// <returns>Output Tensors in a Collection of NamedOnnxValue. User must dispose the output.</returns>
+        public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(
+            IReadOnlyCollection<string> inputNames,
+            IReadOnlyCollection<PinnedOnnxValue> inputValues)
+        {
```

Comment:
This creates one more public Run() method (!) -- if there are too many Run() methods, ultimately a user will be confused about when to use which one 

Not sure how to consider/weigh this new API. It would need to be permanently supported :/  Is the benefit of this new API considerable? -- if yes, then it might be worth the user confusion. If the benefits are not substantial, we should reconsider. Any thoughts?

Reply:
In my opinion, the number of overrides is not a concern. Take the commonly used .NET API `Task.Factory.StartNew()` for example, it has 16 overrides to fit all kinds of requirement from different users. The more important thing is to make sure users can easily understand how to use the method(s).

The override number is big because there is a combination of **input style** (NamedOnnxValue/PinnedOnnxValue) / **output style** (NamedOnnxValue/PinnedOnnxValue/emit for DisposableNamedOnnxValue as return value) / **options** (pass/emit). Users can find this pattern without much effort: passing the input/output/option, and that's it.

Reply:
With `Task.Factory.StartNew` however, there is plenty of documentation, and  legitimate combinations (i.e.  if that override was not present, some users would be blocked). 

Based on the performance numbers, it looks like the new API causes 15% reduction in running time for small models (e.g. MNIST) in a multi-threaded scenario.

If this new API is added, it doesn't unblock any  new user scenarios, but adds  a new function  (similar to the original) which could potentially be faster.


The issue with new API is that they have to be supported permanently -- once released, they cannot be withdrawn. Which is fine, except that the motivation for this API is not at the same level as for new SessionOptions  values (i.e. users will be blocked if C# doesn't expose new SessionOptions values). 

For performance related changes,  is it possible to make changes under-the-hood, instead of creating a new hood?

Reply:
I've thought about this possibility, but the fundamental difference between `NamedOnnxValue` and `PinnedOnnxValue` is, the first one does not hold any native resources and the latter holds a pinned memory (and an underlying native OrtValue). We do not want to modify the existing type (and we should not do), so we need a type to implement `IDisposable` to hold the pinned memory. One cannot be replaced by another because of this difference.

If we had a chance to re-design the whole C# API, I may think about merging the `DisposableNamedOnnxValue` with `PinnedOnnxValue` (having a name field in a reusable onnx value may not be a good idea, but anyway this is another topic). But given the current status, it's hard to reuse the existing type. That is why we have this new class `PinnedOnnxValue`, and with this, all added method overrides may have usage that cannot fulfilled by other overrides.

Reply:
> all added method overrides may have usage that cannot fulfilled by other overrides.

Just for context, is there a particular user scenario that is unblocked by this new API? 

I.e. Under what situation (besides small-model performance gains) would a user be required to use the new API, and the original API would not work?

Reply:
if user's requirement is to reach a high performance target, then it's possible that they have to use the new API because the original API would not be fast enough.

functionally, the original API will always work. There is no scenario that a user have to use the new API to get correct output, because somehow the original API does not support. The purpose of this change is to resolve the latency issue, for this goal, we introduced the new type 'PinnedOnnxValue'. Without using type 'PinnedOnnxValue', users are able to run the model, but may not reach their performance target.